File: web/src/proto.ts
/** Mirrors crates/proto — signaling + input events */

export type Role = "host" | "viewer";

export type SignalMessage =
  | { type: "register_host"; session_id: string; pin: string; device_name?: string }
  | { type: "register_viewer"; session_id: string; pin: string }
  | { type: "registered"; role: Role; session_id: string }
  | { type: "error"; message: string }
  | { type: "offer"; sdp: string }
  | { type: "answer"; sdp: string }
  | {
      type: "ice_candidate";
      candidate: string;
      sdpMid?: string | null;
      sdpMLineIndex?: number | null;
    }
  | { type: "heartbeat" }
  | { type: "pong" }
  | { type: "peer_joined"; role: Role }
  | { type: "peer_left" };

export type InputEvent =
  | { type: "pointer"; x: number; y: number; action: number }
  | { type: "key"; key: string; down: boolean }
  | { type: "wheel"; delta_x: number; delta_y: number };

export function send(ws: WebSocket, msg: SignalMessage) {
  ws.send(JSON.stringify(msg));
}

File: mobile/lib/proto.dart
/// Mirrors crates/proto and web/src/proto.ts

enum Role { host, viewer }

Map<String, dynamic> registerViewer({
  required String sessionId,
  required String pin,
}) =>
    {
      'type': 'register_viewer',
      'session_id': sessionId,
      'pin': pin,
    };

Map<String, dynamic> heartbeat() => {'type': 'heartbeat'};

Map<String, dynamic> answer(String sdp) => {'type': 'answer', 'sdp': sdp};

Map<String, dynamic> iceCandidate({
  required String candidate,
  String? sdpMid,
  int? sdpMLineIndex,
}) =>
    {
      'type': 'ice_candidate',
      'candidate': candidate,
      if (sdpMid != null) 'sdpMid': sdpMid,
      if (sdpMLineIndex != null) 'sdpMLineIndex': sdpMLineIndex,
    };

Map<String, dynamic> pointerEvent({
  required double x,
  required double y,
  required int action,
}) =>
    {
      'type': 'pointer',
      'x': x,
      'y': y,
      'action': action,
    };

Map<String, dynamic> keyEvent({required String key, required bool down}) =>
    {'type': 'key', 'key': key, 'down': down};
